Add option to toggle always showing collision shapes

yahkr opened the pull request:
closes https://github.com/godotengine/godot-proposals/issues/11220

Currently even a simple scene can be visually distracting with a few collision shapes:

This PR adds a setting to only show collision shapes when the node is selected vs all the time.

This is with no object selected
<img src="https://github.com/user-attachments/assets/18ab4ddf-a200-426c-b173-6e2af7608db9" width="512">

With this PR:
<img src="https://github.com/user-attachments/assets/620b5cd0-7370-4550-8144-5e30cb2bcc77" width="512">
<img src="https://github.com/user-attachments/assets/5fedd258-78e6-43fb-b2c4-147587811d3f" width="512">


<img src="https://github.com/user-attachments/assets/a8ff0b42-1c16-4b6b-95f7-9be999eee085" width="512">

<img src="https://github.com/user-attachments/assets/507fc175-dbf4-471b-8e67-c8be1e92016e" width="512">

yahkr committed:
Add option to toggle always showing collision shapes

Calinou commented:
I wonder if this would be better suited to the **View** menu at the top of the 3D editor viewport instead.

yahkr commented:
@Calinou 

So after exploring some I'm running into issues of how to connect that view menu setting to the gizmo logic.  without access to the scene's state I can't look up that property.  One solution I have found is adding a fourth option to the gizmo visibility selector:
![image](https://github.com/user-attachments/assets/163b1c53-cbba-4ec2-845b-421a779095e5)

and then replace this check with the following:
https://github.com/godotengine/godot/blob/0c45ace151f25de2ca54fe7a46b6f077be32ba6f/editor/plugins/node_3d_editor_gizmos.cpp#L1062

```cpp
ref->set_hidden(current_state == HIDDEN || (!ref->is_selected() && current_state == SELECTED));
```

Thoughts vs in the Editor Settings/Editors/3d Gizmos/Gizmo Settings?

yahkr commented:
Sorry was away on holiday, hopefully this is more clear. The issue I was running into regarding putting it in the view menu, is how to access that property. Since it is stored within the editstate file of the scene itself. 

  I can't find a clean way of accessing  `Node3DEditorViewport:: view_menu->get_popup()->is_item_checked(...)` from `CollisionShape3DGizmoPlugin::redraw`

as far as I can see the only route of communication between those settings is via https://github.com/godotengine/godot/blob/893bbdfde8ad1f94fb4e6db246ff7075569396ea/editor/plugins/node_3d_editor_gizmos.cpp#L1046

which is where the visibility selection is checked. 

### Solution 1:
Current PR, adding a new setting within  `editors/3d_gizmos/gizmo_settings/always_show_collision_shapes`
IMO the least bloat and my preferred solution.

### Solution 2:
adding a fourth option to the gizmo visiblity for only visible while selected: 

https://github.com/godotengine/godot/blob/893bbdfde8ad1f94fb4e6db246ff7075569396ea/editor/plugins/node_3d_editor_gizmos.h#L153-L156


### Solution 3:
Exposing something akin to EDITOR_GET for the scenes edit state. or pulling the state with something like this:

```cpp
	Node3DEditorViewport *vp = Node3DEditor::get_singleton()->get_last_used_viewport();
	Dictionary d = vp->get_state();
	if (d.has("always_show_collision_shapes")) {
		show_collision_shapes_always = d["always_show_collision_shapes"];
	}
```

but this feels rather messy

KoBeWi commented:
I think the option name is a bit unclear. "Always show", if you disable it, the shapes are going to show _sometimes_. When exactly?
Reversing it into `show_collision_shapes_only_when_selected`, while more verbose, is also more informative.

yahkr commented:
> show_collision_shapes_only_when_selected

Totally agree with this change. Rebased to efa144396d8003c4b3021bca8242ce5cda4d131f & made this change :)

Calinou commented:
I think the setting should be `true` by default. Otherwise, we're bound to get support requests from users wondering where their collision shapes went 🙂

KoBeWi commented on doc/classes/EditorSettings.xml:
```diff
@@ -492,6 +492,9 @@
 			Size of the disk gizmo displayed when editing [Path3D]'s tilt handles.
 		</member>
+		<member name="editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected" type="bool" setter="" getter="">
+			If [code]true[/code], collision shapes in the 3D editor are always visible. If [code]false[/code], collision shapes are only displayed when selected.
```
The description is reversed.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -796,6 +796,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("editors/3d_gizmos/gizmo_settings/bone_axis_length", (float)0.1);
 	EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "editors/3d_gizmos/gizmo_settings/bone_shape", 1, "Wire,Octahedron");
 	EDITOR_SETTING_USAGE(Variant::FLOAT, PROPERTY_HINT_NONE, "editors/3d_gizmos/gizmo_settings/path3d_tilt_disk_size", 0.8, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_RESTART_IF_CHANGED)
+	EDITOR_SETTING_USAGE(Variant::BOOL, PROPERTY_HINT_NONE, "editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected", false, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_RESTART_IF_CHANGED)
```
Why does it require restart?

yahkr replied:
I couldn't find a good way to update all gizmos on setting change.

KoBeWi commented on editor/plugins/gizmos/collision_shape_3d_gizmo_plugin.cpp:
```diff
@@ -355,6 +356,12 @@ void CollisionShape3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
 		}
 	}
 
+	const bool show_collision_shapes_only_when_selected = EDITOR_GET("editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected");
```
It would be more efficient to cache the value of `show_collision_shapes_only_when_selected` and only update when settings change.

KoBeWi commented:
The shapes are still visible without the gizmo:
![image](https://github.com/user-attachments/assets/39fde5ae-1845-437d-a8b2-cc5f437ddf83)
Though it's probably expected? 
Also you can't select them without a gizmo, but it's probably not a problem.

Calinou commented:
>  Though it's probably expected?

This is likely something that was forgotten during rebase: https://github.com/godotengine/godot/pull/90644

yahkr commented:
> > Though it's probably expected?
> 
> This is likely something that was forgotten during rebase: #90644 and #101810

Thanks! Fixed & pushed

- Fixed docs
- Caching EDITOR_GET
- Hiding debug fills now
- Made it so you dont have to restart editor to apply this change

Calinou reviewed (approved):
Tested locally, it works as expected.

https://github.com/user-attachments/assets/a9e67a87-40e2-4082-872f-0e9137076475

KoBeWi commented on editor/plugins/gizmos/collision_shape_3d_gizmo_plugin.h:
```diff
@@ -42,11 +42,14 @@ class CollisionShape3DGizmoPlugin : public EditorNode3DGizmoPlugin {
 
 	Ref<Gizmo3DHelper> helper;
 
+	static bool show_only_when_selected;
```
```suggestion
	inline static bool show_only_when_selected = false;
```

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -796,6 +796,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("editors/3d_gizmos/gizmo_settings/bone_axis_length", (float)0.1);
 	EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "editors/3d_gizmos/gizmo_settings/bone_shape", 1, "Wire,Octahedron");
 	EDITOR_SETTING_USAGE(Variant::FLOAT, PROPERTY_HINT_NONE, "editors/3d_gizmos/gizmo_settings/path3d_tilt_disk_size", 0.8, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_RESTART_IF_CHANGED)
+	EDITOR_SETTING_USAGE(Variant::BOOL, PROPERTY_HINT_NONE, "editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected", false, "", PROPERTY_USAGE_DEFAULT)
```
```suggestion
	_initial_set("editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected", false)
```

KoBeWi commented on editor/plugins/gizmos/collision_shape_3d_gizmo_plugin.cpp:
```diff
@@ -46,9 +47,13 @@
 CollisionShape3DGizmoPlugin::CollisionShape3DGizmoPlugin() {
 	helper.instantiate();
 
+	set_show_only_when_selected(EDITOR_GET("editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected"));
```
```suggestion
	show_only_when_selected = EDITOR_GET("editors/3d_gizmos/gizmo_settings/show_collision_shapes_only_when_selected");
```

KoBeWi commented on editor/plugins/gizmos/collision_shape_3d_gizmo_plugin.cpp:
```diff
@@ -340,6 +345,10 @@ void CollisionShape3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
 		return;
 	}
 
+	if (!p_gizmo->is_selected() && show_only_when_selected) {
```
```suggestion
	if (show_only_when_selected && !p_gizmo->is_selected()) {
```

Repiteo reviewed (approved):
Looks good! Just needs a rebase, and this should be set!
